### PR TITLE
feat(DEV-11981): invoice Payment & Overdue reminders 🔔

### PR DESCRIPTION
**Pull request by radist2s**
Added functionality to specify _Payment_ and _Overdue_ Reminders during the creation and editing of invoices.

### CODE REVIEW

File: packages/sdk-react/src/mocks/reminders/remindersHandlers.ts [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            message: 'Overdue Reminder not found',
+          },
+        },
+        {
+          status: 404,
+        }
+      );
+    }
+
+    Object.assign(overdueReminderToUpdate, jsonBody);
```
Comment (vkuprin) [inline packages/sdk-react/src/mocks/reminders/remindersHandlers.ts:103]: Any specific reason to use here Object.assign intead of spread operator? {...}

File: packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/CreateReceivables.tsx [hunk 2]
```diff
@@ -204,12 +220,39 @@ const CreateReceivablesBase = (props: InvoiceDetailsCreateProps) => {
...
+
+      <CreateInvoiceReminderDialog
+        open={createReminderDialog.open}
+        reminderType={createReminderDialog.reminderType}
+        onClose={closeCreateReminderDialog}
+        onCreate={({ reminderId, reminderType }) => {
+          if (reminderType === 'payment') {
+            setValue('payment_reminder_id', reminderId);
+          } else if (reminderType === 'overdue') {
+            setValue('overdue_reminder_id', reminderId);
```
Comment (vkuprin) [inline packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/CreateReceivables.tsx:243]: To maintain clean JSX logic, you can refactor the code to move the logic into a separate function

File: packages/sdk-react/src/mocks/reminders/remindersHandlers.ts (hunk 1 above)
Comment (radist2s) [reply packages/sdk-react/src/mocks/reminders/remindersHandlers.ts:103]: Is there a specific advantage of using the spread operator here over `Object.assign` for mocks?

File: packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/components/ReminderForm/OverdueReminderForm.tsx [hunk 3]
```diff
@@ -24,114 +23,167 @@ import {
...
+  reminder,
+  ...props
+}: (
+  | {
+      onCreate?(reminderId: string): void;
+    }
+  | { onUpdate?(reminderId: string): void }
+) &
+  OverdueReminderBasePropsProps & {
+    reminder?: components['schemas']['OverdueReminderResponse'];
```
Comment (vkuprin) [inline packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/components/ReminderForm/OverdueReminderForm.tsx:105]:
```ts
}: (
  | {
      onCreate?(reminderId: string): void;
    }
  | { onUpdate?(reminderId: string): void }
) &
  OverdueReminderBasePropsProps & {
    reminder?: components['schemas']['OverdueReminderResponse'];
  }
```

I would prefer to move this in a separate interface as the amount of types here is quite large

File: packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/CreateReceivables.tsx (hunk 2 above)
Comment (radist2s) [reply packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/CreateReceivables.tsx:243]: Extracting the logic into a separate function and creating additional callbacks increases the number of necessary types that will inevitably be added to the new function. This also, as I mentioned earlier, removes the logic from its context. In my opinion, this is a flawed and counterproductive practice.

File: packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/components/ReminderForm/OverdueReminderForm.tsx [hunk 4]
```diff
@@ -24,114 +23,167 @@ import {
...
   } = api.overdueReminders.getOverdueRemindersId.useQuery(
     {
       path: {
-        overdue_reminder_id: id || '',
+        overdue_reminder_id: reminderId || '',
       },
     },
     {
-      enabled: Boolean(id),
+      enabled: Boolean(reminderId),
```
[the remark this answers is not in the document]
Comment (radist2s) [reply packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/components/ReminderForm/OverdueReminderForm.tsx:64]: First, we don’t have any specific convention for this. Second, I don’t really care. Feel free to change it yourself.

File: packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/components/ReminderForm/OverdueReminderForm.tsx (hunk 3 above)
Comment (radist2s) [reply packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/components/ReminderForm/OverdueReminderForm.tsx:105]: I am really curious, why is this component not even exported?

File: packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/reminderCardTermsHelpers.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
+import { ReactNode } from 'react';
```
Comment (vkuprin) [inline packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/reminderCardTermsHelpers.tsx:1]: Consider importing ReactNode with `import type `

Comment (radist2s) [reply packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/reminderCardTermsHelpers.tsx:1]: Cool, go ahead and fix it!

File: packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/reminderCardTermsHelpers.tsx [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+      reminderTerm: reminder.term_2_reminder,
+      termPeriodName: t(i18n)`2nd discount period`,
+      isDueDate: false,
+    },
+    {
+      reminderTerm: reminder.term_final_reminder,
+      termPeriodName: t(i18n)`Due date`,
+      isDueDate: true,
+    },
+  ].reduce<
```
Comment (vkuprin) [inline packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/reminderCardTermsHelpers.tsx:55]: map might be more readable here, as it is done in previous example, if you are not accumulating values in a complex way there is no need for .reduce

File: packages/sdk-react/src/mocks/reminders/remindersHandlers.ts [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    await delay();
+    return HttpResponse.json(newOverdueReminder);
+  }),
+
+  http.patch<
+    { payment_reminder_id: string },
+    | paths['/payment_reminders/{payment_reminder_id}']['patch']['requestBody']['content']['application/json'],
+    | paths['/payment_reminders/{payment_reminder_id}']['patch']['responses']['200']['content']['application/json']
+    | paths['/payment_reminders/{payment_reminder_id}']['patch']['responses']['404']['content']['application/json']
+  >('*/payment_reminders/:payment_reminder_id', async ({ request, params }) => {
```
Comment (vkuprin) [inline packages/sdk-react/src/mocks/reminders/remindersHandlers.ts:46]: Are this huge types:
```ts
    | paths['/payment_reminders/{payment_reminder_id}']['patch']['requestBody']['content']['application/json'],
    | paths['/payment_reminders/{payment_reminder_id}']['patch']['responses']['200']['content']['application/json']
    | paths['/payment_reminders/{payment_reminder_id}']['patch']['responses']['404']['content']['application/json']
```
 to just cast `*/payment_reminders/:payment_reminder_id'` to follow open api spec?

File: packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/reminderCardTermsHelpers.tsx (hunk 6 above)
Comment (radist2s) [reply packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/reminderCardTermsHelpers.tsx:55]: While reduce is absolutely correct syntax, you should get used to it.

File: packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/components/ReminderForm/OverdueReminderForm.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (vkuprin) [reply packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/components/ReminderForm/OverdueReminderForm.tsx:64]: Well, t's just how it's done in entire project, kinda strange, if it's different is here

File: packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/components/ReminderForm/OverdueReminderForm.tsx (hunk 3 above)
Comment (vkuprin) [reply packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/components/ReminderForm/OverdueReminderForm.tsx:105]: Not sure, I follow your question. I'm asking about types and moving the as separate interface

File: packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/components/ReminderForm/OverdueReminderForm.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (radist2s) [reply packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/components/ReminderForm/OverdueReminderForm.tsx:64]: Feel free to make a proposal with the addition of the ESLint rule.

File: packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/reminderCardTermsHelpers.tsx (hunk 6 above)
Comment (vkuprin) [reply packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/reminderCardTermsHelpers.tsx:55]: I suggest you to actually revisit the code to see if using .map() can improve clarity

File: packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/CreateReceivables.tsx (hunk 2 above)
Comment (vkuprin) [reply packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/CreateReceivables.tsx:243]: The benefits of refactoring code to extract logic into separate functions generally outweigh the drawbacks, even wrapping that in useCallback. It promotes clean, modular, and maintainable code. At the end of the day, it's all about developer preference ...

File: packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/reminderCardTermsHelpers.tsx (hunk 5 above)
Comment (saladNights) [reply packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/reminderCardTermsHelpers.tsx:1]: fixed

Comment (saladNights) [reply packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/reminderCardTermsHelpers.tsx]: fixed

**Review by nekedos — APPROVED** [review]

**Pull request merged**